tests/test_cache.py

```diff
@@ -1,504 +1,372 @@
-        for _ in range(n):
-            r = self.s.get(url, verify=False)
-        self.assertLessEqual(time.time() - t, delay * n / 2)
-
-    def test_from_cache_attribute(self):
-        url = httpbin('get?q=1')
-        self.assertFalse(self.s.get(url).from_cache)
-        self.assertTrue(self.s.get(url).from_cache)
-        self.s.cache.clear()
-        self.assertFalse(self.s.get(url).from_cache)
-
-    def test_gzip_response(self):
-        url = httpbin('gzip')
-        self.assertFalse(self.s.get(url).from_cache)
-        self.assertTrue(self.s.get(url).from_cache)
-
-    def test_close_response(self):
-        for _ in range(3):
-            r = self.s.get(httpbin("get"))
-            r.close()
-
-    def test_get_parameters_normalization(self):
-        url = httpbin("get")
-        params = {"a": "a", "b": ["1", "2", "3"], "c": "4"}
-
-        self.assertFalse(self.s.get(url, params=params).from_cache)
-        r = self.s.get(url, params=params)
-        self.assertTrue(r.from_cache)
-        self.assertEqual(r.json()["args"], params)
-        self.assertFalse(self.s.get(url, params={"a": "b"}).from_cache)
-        self.assertTrue(self.s.get(url, params=sorted(params.items())).from_cache)
-
-        class UserSubclass(dict):
-            def items(self):
-                return sorted(super(UserSubclass, self).items(), reverse=True)
-
-        params["z"] = "5"
-        custom_dict = UserSubclass(params)
-        self.assertFalse(self.s.get(url, params=custom_dict).from_cache)
-        self.assertTrue(self.s.get(url, params=custom_dict).from_cache)
-
-    def test_post_parameters_normalization(self):
-        params = {"a": "a", "b": ["1", "2", "3"], "c": "4"}
-        url = httpbin("post")
-        s = CachedSession(CACHE_NAME, CACHE_BACKEND, allowable_methods=('GET', 'POST'))
-        self.assertFalse(s.post(url, data=params).from_cache)
-        self.assertTrue(s.post(url, data=params).from_cache)
-        self.assertTrue(s.post(url, data=sorted(params.items())).from_cache)
-        self.assertFalse(s.post(url, data=sorted(params.items(), reverse=True)).from_cache)
-
-    def test_stream_requests_support(self):
-        n = 100
-        url = httpbin("stream/%s" % n)
-        r = self.s.get(url, stream=True)
-        first_char = r.raw.read(1)
-        lines = list(r.iter_lines())
-        self.assertTrue(first_char)
-        self.assertEqual(len(lines), n)
+        for i in range(5):
+            r = mock_session.get(httpbin('get'), hooks={hook: hook_func})
+        assert state[hook] == 5
+
+
+def test_normalize_params(mock_session):
+    params = {"a": "a", "b": ["1", "2", "3"], "c": "4"}
+
+    assert mock_session.get(MOCKED_URL, params=params).from_cache is False
+    assert mock_session.get(MOCKED_URL, params=params).from_cache is True
+    assert mock_session.get(MOCKED_URL, params={"a": "b"}).from_cache is False
+    assert mock_session.get(MOCKED_URL, params=dict(sorted(params.items()))).from_cache is True
```

Comment:
Since 3.7, dicts are guaranteed to be in insertion order (https://docs.python.org/3.7/library/stdtypes.html#dict has it under changes), thus this test is essentially equivalent to line 148.
I think a good idea would be to "shuffle" params in line 145 or to reverse the sorted dict:

```
{"b": ["1", "2", "3"], "a": "a", "c": "4"}
dict(sorted(params.items(), reverse=True)))
```

Reply:
Good idea, I just added that.